add make and travis options for building for windows

grounded042 opened the pull request:
This is for #765. It will build executables for windows and in the travis job it will publish the windows and linux binaries with the GitHub release when the build is tagged. Out of the box the standalone binary and the query binary won't work as they need the UI files location specified.

Also, before this is merged someone with access to the repo will have to add a secure key to the travis deploy so the files can be sent along to GitHub.

yurishkuro commented on scripts/travis/build-windows.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
is this script necessary? I.e. why not just call `make build-windows` from .travis?

grounded042 replied:
Yup, I can move it to just running `make build-windows` I was following convention and created a script file. I'll change that up.

jpkrohling replied:
Let's be nice to the next person that will touch this feature and keep it consistent with the other parts of the build :)

grounded042 replied:
I'm confused by the two comments - should I use `make build-windows` or call the script?

jpkrohling replied:
I would say that the script is desirable, for consistency's sake. But @yurishkuro has the final word on this.

yurishkuro replied:
Personally I prefer minimalist approach, of the script doesn't add anything then why have it? 

But since it's here already, it's fine to leave it.

yurishkuro commented on Makefile:
```diff
@@ -147,18 +147,34 @@ build_ui:
 
+.PHONY: build-agent-windows
+build-agent-windows:
+	CGO_ENABLED=0 GOOS=windows installsuffix=cgo go build -o ./cmd/agent/agent-windows.exe $(BUILD_INFO) ./cmd/agent/main.go
```
Any reason not to call this `agent.exe`? Is there any other platform that uses `.exe` extension?

grounded042 replied:
Just following convention since the agent for linux is called `agent-linux`. I'll change it up to be `agent.exe`.

grounded042 replied:
@yurishkuro what if we named the different binaries for windows along the lines of `jaeger.<product>.exe` and dropped the `-windows` on all of them?

yurishkuro replied:
my preference would be to have identical names but group them into dirs/archives by target, e.g.
```
jaeger-1.4.0-darwin-amd64.tar.gz
  - agent
  - collector
  - etc
jaeger-1.4.0-linux-amd64.tar.gz
  - agent
  - collector
  - etc
jaeger-1.4.0-windows-amd64.tar.gz
  - agent.exe
  - collector.exe
  - etc
```

grounded042 replied:
I like it. I'll work on getting that setup @yurishkuro.

yurishkuro commented on .travis.yml:
```diff
@@ -54,9 +57,28 @@ script:
 after_failure:
   - if [ "$CROSSDOCK" == true ]; then make crossdock-logs ; else echo 'skipping crossdock'; fi
+
+deploy:
```
I am not familiar with travis deploy workflow - where do these files end up?  And do we need a different naming conventions? E.g. https://prometheus.io/download/:
```
prometheus-2.2.1.darwin-amd64.tar.gz
prometheus-2.2.1.linux-amd64.tar.gz
prometheus-2.2.1.windows-amd64.tar.gz
```

Also, since these are archive files, do we want to bundle all binaries into one archive?

black-adder replied:
also, we should change these commands to only build and deploy these binaries on landing the PR rather than every commit (unless this already does this)

jpkrohling replied:
> We could also add `branch: master` to further pin down when it deploys.

[I don't think that's what we want](https://docs.travis-ci.com/user/deployment/releases/):

> GitHub Releases uses git tags. If the build commit does not have any tags, one will be created in the form of `untagged-*`, where `*` is a random hex string.

grounded042 replied:
@jpkrohling I think that's only if you have `on.branch: true`. I was thinking that if you had multiple `on` options it would only deploy if all conditions were met. I'll look into it.

grounded042 replied:
I verified `When all conditions specified in the 'on:' section are met, your build will deploy.` So if we have `on.tags: true` and `on.branch: master` it will only deploy when it's a commit on master that has a tag already.

yurishkuro commented:
> Out of the box the standalone binary and the query binary won't work as they need the UI files location specified.

fwiw we should fix it using binary assets

grounded042 commented:
I pushed up some new commits:
- Added a script to package files into archives with a version number like @yurishkuro asked for above.
- Added a darwin build option for travis
- Added a linux build option for travis - this was needed because we don't build all the linux binaries all at once and would make archiving everything together a little complicated

yurishkuro commented on .travis.yml:
```diff
@@ -54,9 +63,34 @@ script:
   - if [ "$HOTROD" == true ]; then bash ./scripts/travis/hotrod-integration-test.sh ; else echo 'skipping hotrod example'; fi
+  - if [ "$LINUX" == true ]; then make build-linux ; else echo 'skipping linux'; fi
+  - if [ "$WINDOWS" == true ]; then make build-windows ; else echo 'skipping windows'; fi
+  - if [ "$DARWIN" == true ]; then make build-darwin ; else echo 'skipping darwin'; fi
```
overall this looks good, but this 3 additions concern me as they are going to significantly increase Travis build times. I'd rather have a single DEPLOY=true step that would build files for all targets, which would be a lot faster since we don't have to wait for Travis VM/container to initialize.

Also, there's now a huge amount of duplication in the make file, I think it could be simplified a lot through variables, similar to the deploy.sh file, e.g.
```
linux-binaries: 
    make build-agent build-collector ... GOOS=darwin

build-agent:
    CGO_ENABLED=0 GOOS=$(GOOS) ...
```

yurishkuro replied:
in fact, DEPLOY=true doesn't even need to be a new matrix entry, we can co-locate it with the DOCKER=true entry (since it requires linux binaries anyway).

grounded042 replied:
Thanks for the input @yurishkuro! I'll have time closer to the middle of this week to make these changes.

grounded042 commented:
@yurishkuro I've made the changes we talked about above - decreased the duplication a good bit. I left `build-all-in-one-linux` in the makefile since it is used for building the all in one image and I didn't want to leak setting the `GOOS` var outside of make.

Is there any other changes needed on this? I think I covered all the feedback. The one thing that will need to be done before the merge is setting the deploy key for travis to push the artifacts to GitHub - this will need to be done by someone who has access to publish release artifacts.
